leveldb_test.go

```diff
@@ -545,6 +626,117 @@ func benchmarkDBGets(b *testing.B, useGetMany bool) {
 	}
 }
 
+func BenchmarkDB_Put(b *testing.B) {
+	itemCounts := []int{1000, 10000, 10000}
+
+	for _, itemCount := range itemCounts {
+		b.Run(fmt.Sprintf("Put_%d", itemCount), func(b *testing.B) {
+			benchmarkDBPut(b, itemCount, func(db *DB, wo *WriteOptions, keys, values [][]byte) error {
+				var lastError error
+
+				for i := range keys {
+					err := db.Put(wo, keys[i], values[i])
+					if err != nil {
+						lastError = err
+					}
+				}
+
+				return lastError
+			})
+		})
+
+		b.Run(fmt.Sprintf("Batch_%d", itemCount), func(b *testing.B) {
+			benchmarkDBPut(b, itemCount, func(db *DB, wo *WriteOptions, keys, values [][]byte) error {
+				wb := NewWriteBatch()
+
+				for i := range keys {
+					wb.Put(keys[i], values[i])
+				}
+
+				err := db.Write(wo, wb)
+
+				wb.Close()
+				return err
+			})
+		})
+
+		b.Run(fmt.Sprintf("PutMany_%d", itemCount), func(b *testing.B) {
+			benchmarkDBPut(b, itemCount, func(db *DB, wo *WriteOptions, keys, values [][]byte) error {
+				return db.PutMany(wo, keys, values)
+			})
+		})
+	}
+}
+
+func benchmarkDBPut(b *testing.B, itemCount int, putFunc func(db *DB, wo *WriteOptions, keys, values [][]byte) error) {
+	b.ReportAllocs()
+
+	dbname := tempDir(b)
+	defer deleteDBDirectory(b, dbname)
+	options := NewOptions()
+	options.SetErrorIfExists(true)
+	options.SetCreateIfMissing(true)
+	_ = DestroyDatabase(dbname, options)
+	db, err := Open(dbname, options)
+	if err != nil {
+		b.Fatalf("Database could not be opened: %v", err)
+	}
+	defer db.Close()
+
+	keys := make([][]byte, itemCount)
+	values := make([][]byte, itemCount)
+
+	h := md5.New()
+
+	var byteCount int64
+
+	for i := 0; i < itemCount; i++ {
+		var value [8]byte
+
+		orgID := rand.Int31()
+
+		var ts [8]byte
+		binary.LittleEndian.PutUint64(ts[:], uint64(time.Now().Unix()))
+
+		var b bytes.Buffer
+		b.WriteString("v1|o:")
+		b.WriteRune(orgID)
+		b.WriteString("|p:")
+		b.WriteRune(rand.Int31())
+		b.WriteString("|t:")
+		b.Write(ts[:])
+
+		h.Reset()
+		hash := h.Sum(b.Bytes())
+
+		var key [20]byte
+
+		binary.LittleEndian.PutUint32(key[0:4], uint32(orgID))
+		copy(key[4:], hash)
+
+		binary.LittleEndian.PutUint64(value[:], uint64(time.Now().UnixNano()))
+
+		keys[i] = key[:]
+		values[i] = value[:]
+
+		byteCount += int64(len(key) + len(value))
+	}
+
+	b.ResetTimer()
+
+	for i := 0; i < b.N; i++ {
+		wo := NewWriteOptions()
+		err := putFunc(db, wo, keys, values)
+		wo.Close()
+
+		if err != nil {
+			b.Fatalf("error running benchmark: %v", err)
+		}
+
+		b.SetBytes(byteCount)
+	}
+}
```

Comment:
So one thing that bit me before (in `contexts-db`) for benchmarking is at the beginning of this benchmark, you put some `defer` s and those will actually be included in the timing of the benchmark since defer execution blocks the return of the function. You need to add a `b.StopTimer()` to actually have real benchmarks ;)